### ISSUE TITLE
ResponseNone Not sending until next call

### ISSUE THREAD

**Issue by seamus65** [keatontaylor/alexa-actions#17]
When testing I have found the the ResponseNone event does not get sent until the skill is triggered again.

This means that I cannot use the ResponseNone in my automations.

Is there any way to get this to send the same as Yes and No?

**Comment by honkerst** [comment]
I've also noticed this behaviour. If I have a ResponseNone automation it doesn't trigger the first time, and in addition the 2nd and 3rd times etc it triggers almost instantly, without waiting for a Yes or a No. 

Other that, it's flippin' awesome! cheers!

**Comment by keatontaylor** [comment]
May not be able to do anything about this unfortunately. It seems amazon can decide not send the skill the session ended message immediately and the way ResponseNone works is that when the skill is sent session end, it will fire the ResponseNone event

**Comment by seamus65** [comment]
Thank You Keaton. Fantastic work by the way.

I have mad a script that you ask Alexa if there are any lights on, she checks and tells you which lights are on then asks if you want to turn all the lights off.

`are_lights_on:
  alias: Are Lights On
  sequence:
  - service: notify.alexa_media
    data_template:
      data:
        type: tts
      target:  "{{ {'Bedroom':'media_player.bedroom_echo_show', 'Kitchen':'media_player.kitchen_show', 'Living Room':'media_player.living_room', 'Seamus':'media_player.seamus', 'Workshop':'media_player.workshop', 'Wallpanel':'media_player.wardhead_wallpanel'} [states.sensor.last_alexa.state] }}"
      message: '{% if (is_state("group.lights", "off")) %}
                  There are no lights on
                {% elif is_state("group.lights", "on") %}
                  Yes, the following lights are on.
                {% set entities = [states.light.bar_light, states.light.big_light, states.light.wall_lights, states.light.the_lamp, states.light.front_hall, states.light.hall_lights, states.light.outside_front, states.light.hyleton, states.light.rgb1, states.light.shelves, states.light.whiteboard, states.light.workshop_lights ] %}
                {% for entity in entities %}
                {% if entity.state == "on" %}
                  {{ entity.name }}.
                {% endif %}
                {% endfor %}
                {% endif %}'
  - condition: state
    entity_id: group.lights
    state: 'on'
  - delay: 00:00:05
  - service: script.activate_alexa_actionable_notification
    data_template:
      text: 'Would you like to turn off all the lights?'
      event_id: 'actionable_notification_are_lights_on'
      alexa_device: "{{ {'Bedroom':'media_player.be`

**Comment by keatontaylor** [comment]
So I did a little more digging on this. According to the amazon skill documentation. 

1. The user says "exit" or "quit".
2. The user does not respond or says something that does not match an intent defined in your voice interface while the device is listening for the user's response.
3. An error occurs.

The session should end if any of the following above are true. What doesn't make sense is that a non-response should indeed cause the ResponseNone to trigger the way I have it programmed, since the RespsoneNone event is trigged when the session ends...

**Comment by keatontaylor** [comment]
@seamus65 @honkerst if you could help me debug this, please add  ` print(handler_input.request_envelope.request.reason)` just below `speak_output = home_assistant_object.post_ha_event("ResponseNone")` That way we can log the reason Alexa thinks the session ended. 

Reproduce the issue, and then click the `Logs: Amazon CloudWatch` in the lower left to view the logs and report back

**Comment by seamus65** [comment]
Here are the logs from two attempts.

The first run returned no event response until the second run was initiated, then the ResponseNone event came through. Also no ResponseNone received from the second run. On a third run the ResponseNone from run two came through, nothing came through from Run 3 and the logs didn't show a SessionEndedReason at all that time.

```

Time (UTC +00:00)
Message
2020-05-11

07:44:25
START RequestId: ca5beb82-b93c-477d-a610-7c09ebd54cd3 Version: 12
START RequestId: ca5beb82-b93c-477d-a610-7c09ebd54cd3 Version: 12

07:44:25
END RequestId: ca5beb82-b93c-477d-a610-7c09ebd54cd3
END RequestId: ca5beb82-b93c-477d-a610-7c09ebd54cd3

07:44:25
REPORT RequestId: ca5beb82-b93c-477d-a610-7c09ebd54cd3 Duration: 541.70 ms Billed Duration: 600 ms Memory Size: 512 MB Max Memory Used: 58 MB Init Duration: 257.44 ms
REPORT RequestId: ca5beb82-b93c-477d-a610-7c09ebd54cd3	Duration: 541.70 ms	Billed Duration: 600 ms	Memory Size: 512 MB	Max Memory Used: 58 MB	Init Duration: 257.44 ms	

07:48:28
START RequestId: 1d38e5a1-5bf1-434f-ab16-311307e837a3 Version: 12
START RequestId: 1d38e5a1-5bf1-434f-ab16-311307e837a3 Version: 12

07:48:29
SessionEndedReason.USER_INITIATED
SessionEndedReason.USER_INITIATED

07:48:29
END RequestId: 1d38e5a1-5bf1-434f-ab16-311307e837a3
END RequestId: 1d38e5a1-5bf1-434f-ab16-311307e837a3

07:48:29
REPORT RequestId: 1d38e5a1-5bf1-434f-ab16-311307e837a3 Duration: 517.30 ms Billed Duration: 600 ms Memory Size: 512 MB Max Memory Used: 59 MB
REPORT RequestId: 1d38e5a1-5bf1-434f-ab16-311307e837a3	Duration: 517.30 ms	Billed Duration: 600 ms	Memory Size: 512 MB	Max Memory Used: 59 MB	
```

**Comment by honkerst** [comment]
Mine looks like this. 

On the Echo show you see the name of the automation that's responding when Alexa asks the actionable question. My test loops 4 times if there I don't respond to Alexa's query. I didn't respond at all to all 4 questions. On Q1 it showed the name of my custom skill, on Q2 and Q3 it showed the name of the ResponseNone automation, and on Q4 it shoed the name of my custom skill again. 

`18:18:46
START RequestId: b15c89f1-b523-4163-8e1d-20dc9c30dd26 Version: 10

18:18:46
/var/task/urllib3/connectionpool.py:1004: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings

18:18:46
InsecureRequestWarning,

18:18:47
END RequestId: b15c89f1-b523-4163-8e1d-20dc9c30dd26

18:18:47
REPORT RequestId: b15c89f1-b523-4163-8e1d-20dc9c30dd26 Duration: 1273.37 ms Billed Duration: 1300 ms Memory Size: 512 MB Max Memory Used: 59 MB Init Duration: 260.11 ms

18:19:14
START RequestId: 991e4c69-f493-4b5f-98d0-57977583471c Version: 10

18:19:15
/var/task/urllib3/connectionpool.py:1004: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings

18:19:15
InsecureRequestWarning,

18:19:15
SessionEndedReason.USER_INITIATED

18:19:15
END RequestId: 991e4c69-f493-4b5f-98d0-57977583471c

18:19:15
REPORT RequestId: 991e4c69-f493-4b5f-98d0-57977583471c Duration: 738.36 ms Billed Duration: 800 ms Memory Size: 512 MB Max Memory Used: 59 MB

18:19:43
START RequestId: 7aa2c067-c29d-46d9-b517-6016e6d56fca Version: 10

18:19:43
/var/task/urllib3/connectionpool.py:1004: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings

18:19:43
InsecureRequestWarning,

18:19:43
END RequestId: 7aa2c067-c29d-46d9-b517-6016e6d56fca

18:19:43
REPORT RequestId: 7aa2c067-c29d-46d9-b517-6016e6d56fca Duration: 710.59 ms Billed Duration: 800 ms Memory Size: 512 MB Max Memory Used: 59 MB`

**Comment by keatontaylor** [comment]
In both your cases, the reason the session ended was due to `USER_INITIATED`, so for some reason Alexa thinks the skill is being interrupted before it can complete.

Just by chance are you triggering the actionable notification to multiple Alexa devices at the same time? Otherwise I am rapidly running out of ideas. 

I've also been unable to reproduce this using the latest lambda_function.py code on the automations I run.

**Comment by seamus65** [comment]
I am only triggering to a single device at the minute.

**Comment by honkerst** [comment]
Yeah I am also only triggering one device. My response automations previously did TTS back to the same Echo but I've just tested again and changed them to TTS to a seperate google mini, but the behaviour on the Echo is the same. (Nothing the first time, 2nd 3rd and 4th times ResponseNone fires before Alexa even finishes asking the question).

**Comment by keatontaylor** [comment]
What is the text you're having Alexa say when she launches the skill?

**Comment by honkerst** [comment]
It's the default question from the example "The front door has been unlocked for 5 minutes, would you like me to lock it?".

**Comment by keatontaylor** [comment]
Do you happen to have follow-up mode turned on for the Alexa device?

**Comment by k7franklin** [comment]
Thanks @keatontaylor for the project.

I`m having problem with working with my Alexa. I dont know if the problem is bcz my echo language is portuguese (brazil) and in my language has no follow-up mode, but the automations or the script doesnt "calls" alexa. She doesnt wake up. 
If i say:  "open custom actions", she says the actionable notify, but she doesnt automatically.

This is compatible with other languages? Is this my problem? My skill are perfectly linked and if i "call" the skill works fine, but the automation cannot make alexa to active ask me...

Thanks in advice.

**Comment by honkerst** [comment]
> Do you happen to have follow-up mode turned on for the Alexa device?

No, I don't have follow-up enabled on any of my Alexa devices.

However while checking my follow-up settings for each device I noticed under the 'Communications' heading it says "Network error - tap to retry" on all of my Echos. Everything else about the Echos seems to work, but from googling this is related to Drop-In feature (which  I have never used). I wonder if that's related to what we're seeing. Currently trying to fix the error and will test again.

**Comment by honkerst** [comment]
Cleared that error by logging out and in again on the Alexa app. Unfortunately no change in the behaviour of ResponseNone.

**Comment by keatontaylor** [comment]
I've given this a bit more of a look and I cannot seem to find the cause. I also cannot reproduce it in my setup, so I am unsure what to do now. 

My theory is that this is an amazon bug that is region specific, or something to do with the skill creation process. I hate to ask this, but would it be possible to recreate the skill and see if the issue still persists?

**Comment by seamus65** [comment]
@keatontaylor I deleted the skill and recreated a new one, same issue remains.
I am not using account linking as I could not get it to work.
I am using nabu casa URL with SSL True.
I also noticed this in the latest code:
![image](https://user-images.githubusercontent.com/35989486/81911303-08fb3c00-95c5-11ea-97c1-f3c2e2c1b28d.png)

**Comment by seamus65** [comment]
As a work-around for the minute I have created an input_boolean which I turn on in the initial automation, then in the ResponseNone automation I have two triggers, ResponseNone and input_boolean on for x seconds (for example), I also turn off the input_boolean with any response received.

**Comment by honkerst** [comment]
This is exactly what I've done as well. It works well enough that I'm not too worried about the ResponseNone problem any more. 

Still loving concept though, dreaming up ways to use this power wisely.

**Comment by renatothome** [comment]
> Thanks @keatontaylor for the project.
> 
> I`m having problem with working with my Alexa. I dont know if the problem is bcz my echo language is portuguese (brazil) and in my language has no follow-up mode, but the automations or the script doesnt "calls" alexa. She doesnt wake up.
> If i say: "open custom actions", she says the actionable notify, but she doesnt automatically.
> 
> This is compatible with other languages? Is this my problem? My skill are perfectly linked and if i "call" the skill works fine, but the automation cannot make alexa to active ask me...
> 
> Thanks in advance.

I was facing the same issue and after updating alexa_media to the latest version available (by now is v2.7.5) it's working fine. I updated it by HACS. 

Good luck!

**Comment by k7franklin** [comment]
> > Thanks @keatontaylor for the project.
> > I`m having problem with working with my Alexa. I dont know if the problem is bcz my echo language is portuguese (brazil) and in my language has no follow-up mode, but the automations or the script doesnt "calls" alexa. She doesnt wake up.
> > If i say: "open custom actions", she says the actionable notify, but she doesnt automatically.
> > This is compatible with other languages? Is this my problem? My skill are perfectly linked and if i "call" the skill works fine, but the automation cannot make alexa to active ask me...
> > Thanks in advance.
> 
> I was facing the same issue and after updating alexa_media to the latest version available (by now is v2.7.5) it's working fine and I'm also using PT-BR. I updated it by HACS.
> 
> Good luck!

Thanks, my friend. This works fine now.

**Comment by keatontaylor** [comment]
I'm going to close this now as it seems the latest version of Alexa Media Player takes care of it :)

**Comment by dudu631** [comment]
Hi, i'm having the exact same issue with the ResponseNone now, and my Alexa Media Player is up to date. 

Can someone help?

I have the account linked, but I did not know which Client Id to use on Brazil, so I used the one for US( https://pitangui.amazon.com/ ). Everything but the ResponseNone, works alright...

**Comment by dudu631** [comment]
I also checked the logs as you suggested above and it's the same issue:
"SessionEndedReason.USER_INITIATED"

**Comment by dudu631** [comment]
Also, i did move my account to Amazon US, and the same behavior occurs. I'm wondering if it's not some region issue with the Echo Device.

**Comment by seamus65** [comment]
Mine still has the issue. It never has given a ResponseNone.
It might start working for me one day.
I am uk based using the eu servers.

**Comment by honkerst** [comment]
Yeah me too :) Last week I finally got round to testing if ResponseNone was fixed, so I reverted my [earlier changes](https://github.com/keatontaylor/alexa-actions/issues/17#issuecomment-628503247) and checked it out. Unfortunately the behaviour seems to be no different. The ResponseNone fires only at the 2nd iteration, and starts doing it before Alexa has even finished asking the question.

**Comment by dudu631** [comment]
So after some time debugging it, I found that in my case is related to the echo show 5. On my echo dot 3rd generation the ResponseNone events are working flawless, when I try the same service call with the same parameters on my Echo Show 5 it breaks: the SessionEndedRequestHandler is only handled on the next subsequent call. Still trying to figure it out why.

**Comment by dudu631** [comment]
I finally figured the cause.

https://forums.developer.amazon.com/questions/202990/sessionendedrequest-not-fired-for-display-timeout.html

This issue describes whats going on. It appears that on Show devices, not answering to the question does not trigger EXCEEDED_MAX_REPROMPTS as the reason of the session ending because the skill remains active on the screen of the device even if there wasn't an answer and then, when the device finally moves to the HomeScreen there is no trigger also.

The "SessionEndedReason.USER_INITIATED" we get on the second call is because you are forcing that session to end, and that's why it is USER_INITIATED. As an improvement, we should check the reason of the SessionEnded as EXCEEDED_MAX_REPROMPTS in order to send the ResponseNone to HomeAssistant.

As for the issue with the show devices, I'll look for a solution that we could do if they don't change this behavior.

**Comment by honkerst** [comment]
That's useful information! Just to confirm I've also been using an Echo Show 5 in my tests.

**Comment by dudu631** [comment]
Great @honkerst.

I asked on the official ask python sdk and amazon developers forums how would I handle the no response situation on display devices. I tried documentation and a lot of searching but no luck so far.

In the meantime, I opened a pull request to prevent the USER_INITIATED wrong event ResponseNone push on Home Assistant. That way, Alexa display devices won't trigger oddly a ResponseNone event, and non-display devices won't trigger ResponseNone if you call it twice consecutively by mistake for some reason.

https://github.com/keatontaylor/alexa-actions/pull/43
